core/src/km_core_state_context_set_if_needed.cpp

```diff
@@ -0,0 +1,220 @@
+ * app context
+ *
+ * TODO: #10100 will alter some of the assumptions here
+ */
+bool is_context_valid(km_core_cp const * new_app_context, km_core_cp const * app_context) {
+  if (new_app_context == nullptr || app_context == nullptr || *app_context == '\0') {
+    // If the app_context is "empty" then it needs updating
+    return false;
+  }
+  km_core_cp const* new_app_context_p = new_app_context;
+  while(*new_app_context_p) {
+    new_app_context_p++;
+  }
+
+  km_core_cp const* app_context_p = app_context;
+  while(*app_context_p) {
+    app_context_p++;
+  }
+
+  // we need to compare from the end of the cached context
+  for(; new_app_context_p >= new_app_context && app_context_p >= app_context; new_app_context_p--, app_context_p--) {
+    if(*new_app_context_p != *app_context_p) {
+      // The cached context doesn't match the application context, so it is
+      // invalid
+      return false;
+    }
+  }
+
+  if(app_context_p > app_context) {
+    // if the cached context is longer than the application context, then we also
+    // assume that it is invalid
+    return false;
+  }
+
+  // It's acceptable for the application context to be longer than the cached
+  // context, so if we match the whole cached context, we can safely return true
+  return true;
+}
+
+/**
+ * Retrieves the context as a km_core_cp string, dropping markers
+ */
+km_core_cp* get_context_as_string(km_core_context *context) {
+  size_t buf_size = 0;
+  km_core_context_item* context_items = nullptr;
+
+  if(km_core_context_get(context, &context_items) != KM_CORE_STATUS_OK) {
+    return nullptr;
+  }
+
+  if(km_core_context_items_to_utf16(context_items, nullptr, &buf_size) != KM_CORE_STATUS_OK) {
+    km_core_context_items_dispose(context_items);
+    return nullptr;
+  }
+
+  km_core_cp *app_context_string = new km_core_cp[buf_size];
+
+  km_core_status status = km_core_context_items_to_utf16(context_items, app_context_string, &buf_size);
+  km_core_context_items_dispose(context_items);
+
+  if(status != KM_CORE_STATUS_OK) {
+    return nullptr;
+  }
+
+  return app_context_string;
+}
+
+/**
+ * Updates the context from the new_context km_core_cp string
+ */
+bool set_context_from_string(km_core_context *context, km_core_cp const *new_context) {
+  km_core_context_item* new_context_items = nullptr;
+
+  km_core_status status = km_core_context_items_from_utf16(new_context, &new_context_items);
+  if (status != KM_CORE_STATUS_OK) {
+    return false;
+  }
+
+  km_core_context_set(context, new_context_items);
+  km_core_context_items_dispose(new_context_items);
+
+  return true;
+}
+
+/**
+ * Normalize the input string using ICU
+ */
+bool do_normalize_nfd(km_core_cp const * src, std::u16string &dst) {
+  UErrorCode icu_status = U_ZERO_ERROR;
+  const icu::Normalizer2 *nfd = icu::Normalizer2::getNFDInstance(icu_status);
+  assert(U_SUCCESS(icu_status));
+  if(!U_SUCCESS(icu_status)) {
+    // TODO: log the failure code
+    return false;
+  }
```

Comment:
could be (with including `ldml_utils.hpp` though in the wrong directory):

```suggestion
  if(!UASSERT_SUCCESS(icu_status)) {
    return false;
  }
```

perhaps it should be hoisted?

Reply:
I looked at doing this originally, but would require some refactoring of the ldml_utils.hpp and cleanup of namespaces -- the macro refers implicitly to code in `km::core::ldml` namespace -- so opted to leave it out for now.